Advanced ptrace

benmcmillanwarwick opened the pull request:
This is a whole bunch of features somewhat clumped together: I could potentially go back and split this up. Partly, just trying to figure out what the best strategy is, and to make sure the design decisions are OK.

Mostly:

1) Utility routines for field evaluation at arbitrary positions.
2) Lorentz timestep modularised to split out charge assignment and field evaluation.
3) Evaluator for spatial derivative of B field (for drift-kinetic solver)
4) Drift-kinetic solver for particle motion.
5) Coupling to 1D Fluid solver.

benmcmillanwarwick committed:
-Mod makefile for gfortran on ubuntu.

-drift-kinetic logical option in particle type.

-Add basics of a drift-kinetic solver. For the moment, this is particle-tracing, and currents of d-k markers are not calculated. Compiles but drift-kinetic option not tested.

benmcmillanwarwick committed:
-Various modifications/new routines to make testing easier.

-Subroutine to allow (particle-weighted) E field and B field and
 B field derivative tensor calculation at arbitrary position.

-Should allow testing of whether the derivative is being calculated
 correctly. Needs some input testing switches.

----------------

-Calculate various terms in drift-kinetic motion evaluation.

-Compiles, runs but not using new code yet.

benmcmillanwarwick committed:
-Warning: epoch doesn't run as usual in this commit.

-Various modifications to test new terms for drift-kinetics.

-Fixed lots of errors in the B field tensor derivatives and standalone field evaluation.

-Initialise a sinusoidal field.

-'fixed_fields' option to prevent time-evolution of fields.

-Call testing routines before main epoch time-loop: needs to be turned off for normal EPOCH execution.

-Testing routines output B field and derivatives at spatial points.

-Checked that dB_x/dx and B_x are consistent (this should be machine-precision due to the
 finite-element-like approach)

-Needs more testing.

benmcmillanwarwick committed:
-Fix bug in x component of B field derivative

-Modified testing setup.

benmcmillanwarwick committed:
-Basic Drift-Kinetic particle tracing looks roughly right: not systematically tested.
   -Set up for lowest-order drift-kinetic treatment in static fields.

-Setup loads a bunch of electrons with constant temp/density as drift-kinetic particles.

-Remove pre-epoch test.

-Transform velocity coordinates to drift-kinetic after load.

-Various bug fixes in drift-kinetic tracing routines.

-

benmcmillanwarwick committed:
-Second order drift-kinetic solve. Visually OK.

-Pull field evaluation out of drift-kinetic timestep.

benmcmillanwarwick committed:
-Because we need B field to transform to drift-kinetic coordinates, new dependency in helper.f90

-The fields evaluator ought to be moved out of the particles file, into fields.f90.

benmcmillanwarwick committed:
-Split out current deposition from push. (subroutine current_deposition)

-Added a new push routine using split fields evaluation and current deposition.

benmcmillanwarwick committed:
-Remove local particle coordinate from split push

-Fix bugs in positions in split push

benmcmillanwarwick committed:
- Add helper function to allow direct call of current deposition given
  macroparticle charge and start and end of the line on which the particle
  moves.

- Flip arguments so current deposition arguments are in order
  (from location, to location).

- Some cleanup.

benmcmillanwarwick committed:
-Basic tests: split timestep gives almost exactly same results for two-stream as original code although some differences at
 later time near machine precision.

-Fix some bugs: e.g. field quantities include the partition-of-unity coming from the splines.
                e.g. needed to zero some variables explicitly.

-Some unnecessary quantities were being calculated that are only needed for drift-kinetic solve; isolate these so they are
 not calculated unless needed.

benmcmillanwarwick committed:
-Various major/minor modifications to support a second-order drift-kinetic solve.

-Need temporary storage for particle quantities: work array in particles which means
 basic PIC particle push is currently very memory inefficient

-Split out the drift-kinetic and standard-PIC currents.

-Assign the current from d-k particles (just due to marker motion, not magnetisation
 current yet).

-Also temporary storage for field arrays: need to 'rewind' the field by half a
 timestep to allow a centred evaluation for the drift-kinetic solve, and so the
 final 2nd-order accurate Maxwell solve can proceed.

-The 2nd order dk push is called after the end of the main leapfrog push. d-k particle
 phase-space positions are stored on the half-timestep.

benmcmillanwarwick committed:
-Add back particle output for diagnosis

-Changed comment

benmcmillanwarwick committed:
-Add deltaf to minepoch | Add (1D) fluid solver to minepoch

-Added pvol attribute in particles. (for deltaf)

-Various setup for deltaf, initialisation of pvol from main EPOCH

-Fluid solver is implemented in particles.F90. Simple lowest-order two-moment 1D upwind.

-deltaf is calculated using time-evolved fluid background.

-Added 'one-stream' case with a cold-ish plasma to compare against fluid.
 E=B=0 fixed for testing free-streaming cold-fluid case.

benmcmillanwarwick committed:
Merge branch 'main' into advanced_ptrace

benmcmillanwarwick committed:
Merge remote-tracking branch 'remotes/origin/tom/input_deck' into advanced_ptrace

Added a select_case statement instead of IF statements for 'problem'.

benmcmillanwarwick committed:
-Cleanup of fluid solver case so these routines are off by default

-Preparatory to pull back into master

TomGoffrey commented:
I think `fixed_fields` should be added to the control namelist of the input deck. I would also think it ought to be initialised in `control_initialise` rather than `fields_initialise`, as well as (possibly) having problem dependent defaults (it should be `.FALSE.` for the two-steam instability, for example). Arguably the global default should be `.FALSE.`.

However, even re-enabling the field solver by setting `fixed_fields = .FALSE.` in `fields_initialise` doesn't seem to re-enable the solver?

benmcmillanwarwick committed:
-Add fixed_fields to deck.

-fixed_fields default to .FALSE.

-Remove zeroing of E,B in set_initial_values: this is done in fields_initialise.

-one_stream_setup needed to follow same logic as other custom problems.

benmcmillanwarwick committed:
-Move output after final Runge-Kutta substep.

Merge remote-tracking branch 'remotes/origin/main' into advanced_ptrace

benmcmillanwarwick committed:
-Bug fix: switch to put current in drift-kinetic vs. normal currents in current deposition
 was wrong way around.

-Make the force_write switch do something in diagnostics (is this the right thing though?)

benmcmillanwarwick commented:
OK, there was at least one significant bug in the current deposition. At least there is some field energy being generated now.

I think fixed_fields is now set up in a sensible way.

benmcmillanwarwick committed:
Merge remote-tracking branch 'origin/main' into advanced_ptrace

TomGoffrey commented:
This looks good now. Matches the growth rates observed on main.

For example, on 4 cores:

![two_stream](https://user-images.githubusercontent.com/50208076/108629416-ab4f2680-7457-11eb-97a9-775638448b0f.png)

I should say, the code which calculates the growth rates is neither intelligent, nor robust. I get answers of 0.7247 and 0.7265 on 1 and 2 cores, for example.

I haven't checked the correctness of the new features you've added. Let me know if you want me to add CI runs with them and I can add it to my next PR.

TomGoffrey commented on src/epoch3d.F90:
```diff
@@ -136,10 +139,20 @@ PROGRAM pic
 #endif
 
     step = step + 1
-    time = time + dt
```
Is this repositioning of the time increment deliberate?

For core EPOCH (with time-dependent EM boundary conditions), the time increase `time = time + dt` must happen before `CALL update_eb_fields_final` to enable the correct boundary condition to be applied.

benmcmillanwarwick replied:
Sorry, wasn't being very careful.

I've fiddled with it now, but I think each of the different sub-stepping methods might benefit from maintaining their own internal concept of where in the timestep they are, rather than having it as a global.

TomGoffrey replied:
Yeah. I couldn't work out what the correct solution was for the different solver options either. Agree it's worth considering each solver maintaining their own step.

benmcmillanwarwick committed:
-Move time variable update back before fields_update_final
